docs/root/configuration/listeners/listener_filters/original_dst_filter.rst

```diff
@@ -40,5 +40,26 @@ rather than the address at which the listener is listening at. Furthermore, :ref
 destination cluster <arch_overview_service_discovery_types_original_destination>` may be used to
 forward HTTP requests or TCP connections to the restored destination address.
 
+Internal listeners
+------------------
+
+Original destination listener filter reads the dynamic metadata and the filter
+state objects on the user space sockets handled by the :ref:`internal listeners
+<config_internal_listener>`.
+
+Dynamic metadata for the destination address is expected to be placed into the
+key `envoy.filters.listener.original_dst` under the field `local` and should
+contain a string with an IP and a port address.
```

Comment:
Looking at the existing use of original_dst and this form, it seems the contract with the rest of Envoy is different. This new use case is about populating filter state, rather than original_dst cluster. Feels a bit overloaded, should there be some plans to separate source of information in some consistent metadata from consumption?

Reply:
Are you thinking of a "common" metadata for the destination? This change is about consuming a filter state and metadata, so the typical chain is the following:

1. Some listener/cluster that produces filter state and/or metadata to an internal address.
2. Internal listener with original_dst listener filter.
3. Any cluster, common to have original_dst cluster here.

I was thinking that original_dst listener filter updates the addresses, we no longer need to modify original_dst cluster. But it already has a filter state defined for the same purpose, so we could reuse the same object. This approach might interfere with normal original_dst cluster behavior because filter state takes priority and there is no way to delete a filter state object.

Reply:
Are you saying we're using the same filter state as the existing original_dst cluster? If so, I suggest refactoring the docs to make this explicit.

Reply:
No, we're using different filter states. ORIGINAL_DST cluster uses [envoy.network.transport_socket.original_dst_address](https://www.envoyproxy.io/docs/envoy/latest/configuration/advanced/well_known_filter_state#well-known-filter-state). original_dst listener filter uses `envoy.filters.listener.original_dst.local_ip`. I thought you suggested to share them? I think it's safer not to share.

Reply:
I did a bit of refactoring to make it clear that ORIGINAL_DST cluster and filter share the object but have different keys.